### PR TITLE
[SG-658] Refactor reactive form logic

### PR DESCRIPTION
**Pull request by gbubemismith**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Refactor the Login` form to prevent the required email validation error from showing up when `create account` is clicked and refactor the reactive form logic to only show the error after typing has happened on a field. 

## Code changes

- **apps/web/src/app/accounts/login/login.component.html:** Used `mouseDown` event to prevent onblur from showing  validation error when the email field is cleared and then `create account` is clicked.
- **apps/web/src/app/accounts/login/login.component.ts** Added `goToRegister` method to route to the registration form.
- **libs/components/src/input/input.directive.ts** Modified input directive to only show validation error when typing has been completed on a field

## Screenshots

https://user-images.githubusercontent.com/13024008/210433829-bb28f8bb-80ba-424d-938c-4ebcb0f90637.mov

https://user-images.githubusercontent.com/13024008/210636743-b6fa0a4b-cc98-411b-92a9-db749310eb86.mov

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: apps/web/src/app/accounts/login/login.component.html [hunk 1]
```diff
@@ -63,7 +63,7 @@
 
           <p class="tw-m-0 tw-text-sm">
             {{ "newAroundHere" | i18n }}
-            <a routerLink="/register">{{ "createAccount" | i18n }}</a>
+            <a [routerLink]="[]" (mousedown)="goToRegister()">{{ "createAccount" | i18n }}</a>
```
Comment (JaredSnider-Bitwarden) [inline apps/web/src/app/accounts/login/login.component.html:66]: I tested this locally and they both work. 

However, I would say that `(click)` is a bit more of a standard event handler for this purpose.  

Just as info, you can achieve the same result (getting the styling w/out breaking routing) as `[routerLink]="[]"` by using just `routerLink` or `routerLink=""`

File: libs/components/src/input/input.directive.ts [hunk 2]
```diff
@@ -63,12 +72,33 @@ export class BitInputDirective implements BitFormFieldControl {
   @Input() hasPrefix = false;
   @Input() hasSuffix = false;
 
+  @Input() showErrorsWhenDisabled? = false;
```
Comment (JaredSnider-Bitwarden) [inline libs/components/src/input/input.directive.ts:75]: This looks like the same logic that I added in this PR to the [domain verification feature branch](https://github.com/bitwarden/clients/blob/feature/web-organization-domain-claiming/libs/components/src/input/input.directive.ts). Why did you need that logic here?  The `showErrorsWhenDisabled` property isn't used on any inputs in your branch? Sorry if I'm missing something.

**Comment by JaredSnider-Bitwarden** [comment]
Overall, this is looking awesome. Great work on this. I also pulled it down and tested out the logic to carry over the email from the login page to the create account form. 
https://user-images.githubusercontent.com/116684653/210446354-a0b5c8cc-9111-489d-8855-da0d27af34f4.mov

File: libs/components/src/input/input.directive.ts (hunk 2 above)
Comment (gbubemismith) [reply libs/components/src/input/input.directive.ts:75]: Yeah, I added your logic so we wouldn't have conflicts in the future, but I think it's best I remove that part

File: apps/web/src/app/accounts/login/login.component.html (hunk 1 above)
Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/accounts/login/login.component.html:66]: After a discussion, `click` cannot be used as it causes the form validation to fire and move the anchor link down below so the user's click on create account doesn't work as the click event never fires for the link. `mousedown` does not trigger form validation so it is a clever fix.

**Review by JaredSnider-Bitwarden — COMMENTED** [review]
Good work! Will approve pending design review.

**Review by sukhleenb — APPROVED** [review]

**Review by Hinton — COMMENTED** [review]
Unfortunately this seems to break the expected form behavior. Causing errors to not appear when clicking submit, see https://62a88a6de5b807fa98886113-vqzexjldcq.chromatic.com/?path=/story/component-library-form-error-summary--default

A possible approach is to have no validation on the login form until you click the button (which could be combined with a `onBlur` on the input). I think in general it would be interesting to investigate if we could avoid triggering validations if the app router is transitioning but I suspect that would involve hooking into the router.

If we want to change the generic CL please verify the existing forms behavior isn't impacted.

**Comment by gbubemismith** [comment]
The expectation on this ticket had to be modified a bit. Changing `this.ngControl?.touched` to `this.ngControl?.dirty` affected the default form behaviour where field errors were not displayed when a form was submitted. So we decided to stick with the default behaviour and add an extra behaviour not to show an error message on a field while typing.

File: apps/web/src/app/accounts/login/login.component.html [hunk 3]
```diff
@@ -63,7 +63,8 @@
 
           <p class="tw-m-0 tw-text-sm">
             {{ "newAroundHere" | i18n }}
-            <a routerLink="/register">{{ "createAccount" | i18n }}</a>
+            <!--mousedown event is used over click because it prevents the validation from firing  -->
+            <a [routerLink]="[]" (mousedown)="goToRegister()">{{ "createAccount" | i18n }}</a>
```
Comment (Hinton) [inline apps/web/src/app/accounts/login/login.component.html:67]: Is it possible  to keep the `register` portion here? It acts as a hint to the user that they will navigate to register, if not we should change it to a button.

**Comment by trmartin4** [comment]
When this is merged, it will also address Github issue https://github.com/bitwarden/clients/issues/4404.

File: apps/web/src/app/accounts/login/login.component.html (hunk 3 above)
Comment (gbubemismith) [reply apps/web/src/app/accounts/login/login.component.html]: `Register` on the `routerLink` won't affect the behaviour and `mousedown` works on touch devices

**Review by danielleflinn — APPROVED** [review]
